src/KafkaNetClient/KafkaTcpSocket.cs

```diff
@@ -138,92 +142,112 @@ private Task<byte[]> EnqueueReadTask(int readSize, CancellationToken cancellatio
-                    }
-
-                    if (ex is ServerDisconnectedException)
-                    {
-                        if (OnServerDisconnected != null) OnServerDisconnected();
-                        _log.ErrorFormat(ex.Message);
-                        continue;
-                    }
-
+                    //it handel the Pennding task inside so we don't need to SetDisposeExceptonToPenndingTasks
                     _log.ErrorFormat("Exception occured in Socket handler task.  Exception: {0}", ex);
+                    if (OnServerDisconnected != null) OnServerDisconnected();
                 }
             }
         }
 
-        private void SetDisposeExceptonToPenndingTask()
+        private void SetDisposeExceptonToPenndingTasks(Exception exception = null)
         {
-            _log.WarnFormat("KafkaTcpSocket thread shutting down because of a dispose call.");
             var disposeException = new ObjectDisposedException("Object is disposing.");
-            _sendTaskQueue.DrainAndApply(t => t.Tcp.TrySetException(disposeException));
-            _readTaskQueue.DrainAndApply(t => t.Tcp.TrySetException(disposeException));
+            if (exception == null)
+            {
+                _log.WarnFormat("KafkaTcpSocket thread shutting down because of a dispose call.");
+                _sendTaskQueue.DrainAndApply(t => t.Tcp.TrySetException(disposeException));
+                _readTaskQueue.DrainAndApply(t => t.Tcp.TrySetException(disposeException));
+            }
+            else
+            {
+                _log.WarnFormat("KafkaTcpSocket not able to connect cancel all PenndingTasks.");
+                _sendTaskQueue.DrainAndApply(t => t.Tcp.TrySetException(exception));
+                _readTaskQueue.DrainAndApply(t => t.Tcp.TrySetException(exception));
+            }
         }
 
         private async Task ProcessNetworkstreamTasks(NetworkStream netStream)
         {
-            Task lastWriteTask = Task.FromResult(true);
-            Task lastReadTask = Task.FromResult(true);
-
             //reading/writing from network steam is not thread safe
             //Read and write operations can be performed simultaneously on an instance of the NetworkStream class without the need for synchronization.
             //As long as there is one unique thread for the write operations and one unique thread for the read operations, there will be no cross-interference
             //between read and write threads and no synchronization is required.
             //https://msdn.microsoft.com/en-us/library/z2xae4f4.aspx
+
+            //Exception need to thrown immediately and not depend on the next task
+            var readTask = ProcessNetworkstreamsSendTask(netStream);
+            var sendTask = ProcessNetworkstreamTasksReadTask(netStream);
+            await Task.WhenAny(readTask, sendTask);
+            if (_disposeToken.IsCancellationRequested) return;
+            await ThrowTaskException(readTask);
```

Comment:
Maybe change the function name to something that implies that an exception will not be necessarily thrown. Maybe ThrowTaskExceptionIfFaulted